Location lookup table

jioter opened the pull request:
Location lookup table

jioter committed:
Fix Validation usage

jioter committed:
Modified Location entity
Add Country entity
Changes to Location to liquibase
Add Country by liquibase
Insert data into location and country

jioter committed:
Modified Location entity
Add Country entity
Changes to Location to liquibase
Add Country by liquibase
Insert data into location and country
Add CountryResponseVO
Add CountryMapper
LocationVO moved to package

jioter committed:
Modified Location entity
Add Country entity
Changes to Location to liquibase
Add Country by liquibase
Insert data into location and country
Add CountryResponseVO
Add CountryMapper
LocationVO moved to package

jioter committed:
Modified Location entity
Add Country entity
Changes to Location to liquibase
Add Country by liquibase
Insert data into location and country
Add CountryResponseVO
Add CountryMapper
LocationVO moved to package

jioter committed:
Modified Location entity
Add Country entity
Changes to Location to liquibase
Add Country by liquibase
Insert data into location and country
Add CountryResponseVO
Add CountryMapper
LocationVO moved to package
Fixed response for mapper CountryCode

jioter committed:
Modified Location entity
                                     Add Country entity
                                     Changes to Location to liquibase
                                     Add Country by liquibase
                                     Insert data into location and country
                                     Add CountryResponseVO
                                     Add CountryMapper
                                     LocationVO moved to package
                                     Fixed response for mapper CountryCode
                                     Removed unused code

jioter committed:
Modified Location entity
                                     Add Country entity
                                     Changes to Location to liquibase
                                     Add Country by liquibase
                                     Insert data into location and country
                                     Add CountryResponseVO
                                     Add CountryMapper
                                     LocationVO moved to package
                                     Fixed response for mapper CountryCode
                                     Removed unused code
                                     Fixed some tests
                                     TEST - shouldCreateLocation - still doesn't work

jioter committed:
Modified Location entity
                                     Add Country entity
                                     Changes to Location to liquibase
                                     Add Country by liquibase
                                     Insert data into location and country
                                     Add CountryResponseVO
                                     Add CountryMapper
                                     LocationVO moved to package
                                     Fixed response for mapper CountryCode
                                     Removed unused code
                                     Fixed some tests
                                     TEST - shouldCreateLocation - still doesn't work

Add MVC find Cities by CountryCode

jioter committed:
Modified Location entity
                                     Add Country entity
                                     Changes to Location to liquibase
                                     Add Country by liquibase
                                     Insert data into location and country
                                     Add CountryResponseVO
                                     Add CountryMapper
                                     LocationVO moved to package
                                     Fixed response for mapper CountryCode
                                     Removed unused code
                                     Fixed some tests
                                     TEST - shouldCreateLocation - still doesn't work

Add MVC find Cities by CountryCode

Add CityMapper

jioter committed:
Merge branch 'develop' of https://github.com/dmitriychernov7/sandbox-team3-be into location-lookup-table

dmitriychernov7 commented on src/main/java/com/exadel/discountwebapp/location/controller/LocationController.java:
```diff
@@ -29,8 +34,23 @@ public LocationResponseVO getLocationById(@PathVariable Long id) {
 
     @GetMapping("/country")
     public List<LocationResponseVO> getAllLocationsByCountry(
-            @RequestParam(name = "country") String country) {
-        return locationService.findAllByCountry(country);
+            @RequestParam(name = "countryCode") String countryCode) {
+        return locationService.findAllByCountryCode(countryCode);
+    }
```
In the future this endpoint should be replaced with getLocation with rsql query parameter

as well as getAllLocationsByCity

jioter replied:
Okay. Will keep in mind this.

dmitriychernov7 commented on src/main/java/com/exadel/discountwebapp/location/entity/Country.java:
```diff
@@ -0,0 +1,31 @@
+    @Column(name = "country_full_name", length = 255, nullable = false, unique = true)
+    String countryFullName;
+
+    @ToStringExclude
+    @OneToMany(mappedBy = "countryCode")
+    private List<Location> locations;
```
Lookup table should have only dictionary information. It should not has any of business logic. 

It is better to remove this field and make location -> country relationship unidirectional

jioter replied:
Removed

dmitriychernov7 commented on src/main/java/com/exadel/discountwebapp/location/entity/Location.java:
```diff
@@ -27,12 +25,17 @@
+    @Column(name = "loc_city", length = 50, nullable = true)
     private String city;
 
+    @ToString.Exclude
```
don't exclude this field from toString

jioter replied:
Removed

dmitriychernov7 commented on src/main/java/com/exadel/discountwebapp/location/entity/Location.java:
```diff
@@ -27,12 +25,17 @@
+    @ToString.Exclude
+    @ManyToOne
+    @JoinColumn(name = "country_code")
+    private Country countryCode;
```
rename field to country

jioter replied:
Renamed

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/discount/mapper/DiscountMapper.java:
```diff
@@ -10,10 +10,11 @@
 import com.exadel.discountwebapp.location.entity.Location;
```
Let's remove CategoryValidator.java file if it is empty

jioter replied:
Removed

dmitriychernov7 commented on src/main/java/com/exadel/discountwebapp/location/repository/LocationRepository.java:
```diff
@@ -1,15 +1,28 @@
 public interface LocationRepository extends CrudRepository<Location, Long> {
 
-    List<Location> findAllByCountry(String country);
+    @Query(value =
+            "select location.loc_id, " +
+                    "location.country_code, " +
+                    "location.loc_city, " +
+                    "location.loc_address_line," +
+                    "location.loc_created, " +
+                    "location.loc_modified " +
+                    "from location " +
+                    "left join country " +
+                    "on location.country_code = country.country_code " +
+                    "where country.country_code = :countryCode", nativeQuery = true)
+    List<Location> findAllByCountryCode(@Param("countryCode") String countryCode);
```
can be replaced with 

List<Location> findAllByCountryCode_CountryCode(String countryCode);

dmitriychernov7 replied:
"_" is used to separate entities. like "." in java

jioter replied:
I have tried, previously. Somehow does't work. Throws exception that CountryCode shound be a Country class. If use JPQL with class entities - the same result.

jioter replied:
Error like this:

Traversed path: Location.country.; nested exception is java.lang.IllegalArgumentException: Failed to create query for method public abstract java.util.List com.exadel.discountwebapp.location.repository.LocationRepository.findAllByCountryCode_CountryCode(java.lang.String)! No property code found for type Country! Traversed path: Location.country.

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/discount/mapper/DiscountMapper.java:
```diff
@@ -73,9 +74,11 @@ public DiscountResponseVO toVO(Discount discount) {
-        List<LocationResponseVO> locations = discount.getLocations()
-                .stream().map(locationMapper::toVO)
-                .collect(Collectors.toList());
+        List<LocationResponseVO> locations = new ArrayList<>();
```
Is this approach with for-each iteration better than was with streams? Just curious

jioter replied:
Changed to Functional Style

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/controller/LocationController.java:
```diff
@@ -29,8 +34,23 @@ public LocationResponseVO getLocationById(@PathVariable Long id) {
+    }
+
+    // TODO: tests for this controller
+    @PreAuthorize("isAuthenticated()")
```
we can remove @PreAuthorize("isAuthenticated()")
same below

jioter replied:
Removed

dmitriychernov7 commented on src/main/java/com/exadel/discountwebapp/location/service/CountryService.java:
```diff
@@ -0,0 +1,25 @@
+    private final CountryMapper countryMapper;
+
+    public List<CountryResponseVO> findAllCountries() {
+        List<Country> countries = countryRepository.findAll();
+        List<CountryResponseVO> response = new ArrayList<>();
+        countries.forEach(entity -> response.add(countryMapper.toVO(entity)));
+        return response;
```
this can be replaced with stream (but this is just a suggection, current code also works fine)

return countryRepository.findAll().stream()
                         .map( countryMapper::toVO )
                         .collect( Collectors.toList() );

jioter replied:
Changed to Functional Style

dmitriychernov7 commented on src/main/java/com/exadel/discountwebapp/location/vo/city/CityResponseVO.java:
```diff
@@ -0,0 +1,19 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class CityResponseVO {
+    @NotBlank
+    @Size(min = 2, max = 50)
```
There is no need to have validation annotations in response VOs

They won't be used anywhere

jioter replied:
Removed

dmitriychernov7 commented on src/main/java/com/exadel/discountwebapp/location/vo/country/CountryResponseVO.java:
```diff
@@ -0,0 +1,23 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class CountryResponseVO {
+    @NotBlank
+    @Size(min = 2, max = 50)
```
not needed

jioter replied:
Removed

dmitriychernov7 commented on src/main/java/com/exadel/discountwebapp/location/vo/country/CountryResponseVO.java:
```diff
@@ -0,0 +1,23 @@
+    @Size(min = 2, max = 50)
+    String countryCode;
+
+    @NotBlank
+    @Size(min = 2, max = 255)
```
not needed

jioter replied:
Removed

dmitriychernov7 commented on src/main/resources/db/changelog/init/01072021_create_country_table.yml:
```diff
@@ -0,0 +1,23 @@
+            columns:
+              - column:
+                  name: country_code
+                  type: VARCHAR(50)
```
length can be 3 chars

jioter replied:
Changed to 2

dmitriychernov7 commented on src/main/resources/db/changelog/init/15062021_create_location_table.yml:
```diff
@@ -1,7 +1,7 @@
   - changeSet:
       id: 15062021_create_location_table
-      author: volodymyr.bushko@gmail.com
+      author: volodymyr.bushko@gmail.com & hreshchuk@gmail.com
```
don't need to add new author
initial is enough. all other changes can be viewed via Git

jioter replied:
Removed hresh...

dmitriychernov7 commented on src/main/resources/db/changelog/init/15062021_create_location_table.yml:
```diff
@@ -16,15 +16,22 @@ databaseChangeLog:
-                  name: loc_country
-                  type: VARCHAR(50)
+                  name: country_code
+                  type: VARCHAR(255)
```
length should be the same  as country_code in country lookup table

jioter replied:
Changed

dmitriychernov7 commented on src/main/resources/db/changelog/init/15062021_create_location_table.yml:
```diff
@@ -16,15 +16,22 @@ databaseChangeLog:
                   name: loc_city
                   type: VARCHAR(50)
                   constraints:
                     nullable: false
```
city is nullable now

jioter replied:
Changed

dmitriychernov7 commented on src/test/resources/sql/clean-up.sql:
```diff
@@ -1,4 +1,5 @@
 SET FOREIGN_KEY_CHECKS = 0;
+TRUNCATE TABLE country;
```
this is a lookup table it won't be changed during tests 

it should not be deleted, because after it is deleted all initial data will be removed, but it is required by application

jioter replied:
Removed this line

dmitriychernov7 commented on src/test/resources/sql/location-init.sql:
```diff
@@ -1,9 +1,13 @@
+INSERT INTO  country
+VALUES ('UA', 'Ukraine');
```
country lookup table should not be modified in tests

jioter replied:
Removed

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/entity/Country.java:
```diff
@@ -0,0 +1,31 @@
+    @Column(name = "country_code", length = 50, nullable = false, unique = true)
+    String countryCode;
+
+    @Column(name = "country_full_name", length = 255, nullable = false, unique = true)
```
There is no need in creating unique constraint (index) for the countryFullName field

jioter replied:
Removed

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/entity/Country.java:
```diff
@@ -0,0 +1,31 @@
+public class Country {
+    @EqualsAndHashCode.Exclude
+    @Id
+    @Column(name = "country_code", length = 50, nullable = false, unique = true)
```
no need to specify unique=true since this field is a primary key

moreover standard countryCode length cannot be more than 2 characters

jioter replied:
Chenged to 2

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/entity/Country.java:
```diff
@@ -0,0 +1,31 @@
+    @EqualsAndHashCode.Exclude
+    @Id
+    @Column(name = "country_code", length = 50, nullable = false, unique = true)
+    String countryCode;
```
private?

jioter replied:
Added

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/entity/Country.java:
```diff
@@ -0,0 +1,31 @@
+    String countryCode;
+
+    @Column(name = "country_full_name", length = 255, nullable = false, unique = true)
+    String countryFullName;
```
private

jioter replied:
added

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/entity/Country.java:
```diff
@@ -0,0 +1,31 @@
+@Table(name = "country")
+@EntityListeners(AuditingEntityListener.class)
+public class Country {
+    @EqualsAndHashCode.Exclude
```
let's include countryCode in equals/hashCode, because in this case our primary key is a business key also

jioter replied:
Included

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/mapper/CityMapper.java:
```diff
@@ -0,0 +1,26 @@
+public class CityMapper {
+    private final ModelMapper modelMapper = new ModelMapper();
+
+    @Autowired
```
@Autowired is redundant here

jioter replied:
Removeed

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/vo/country/CountryResponseVO.java:
```diff
@@ -0,0 +1,23 @@
+public class CountryResponseVO {
+    @NotBlank
+    @Size(min = 2, max = 50)
+    String countryCode;
```
private

jioter replied:
added

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/mapper/CountryMapper.java:
```diff
@@ -0,0 +1,26 @@
+    private final ModelMapper modelMapper = new ModelMapper();
+
+    @Autowired
+    public CountryMapper() {
```
@Autowired is redundant here

jioter replied:
Removed

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/service/CountryService.java:
```diff
@@ -0,0 +1,25 @@
+    private final CountryRepository countryRepository;
+    private final CountryMapper countryMapper;
+
+    public List<CountryResponseVO> findAllCountries() {
```
@Transactional

jioter replied:
Added

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/service/LocationService.java:
```diff
@@ -48,14 +51,21 @@ public Location findEntityById(Long id) {
         return getLocationResponseVO(locations);
     }
 
+    public List<CityResponseVO> findAllCitiesByCountryCode(String countryCode) {
```
@transactional

jioter replied:
Added

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/vendor/vo/VendorResponseVO.java:
```diff
@@ -1,6 +1,6 @@
 package com.exadel.discountwebapp.vendor.vo;
```
Let's remove VendorValidator.java empty file for now

jioter replied:
Removed

ahlukhouski2 commented on src/main/resources/db/changelog/init/01072021_create_country_table.yml:
```diff
@@ -0,0 +1,23 @@
+                  type: VARCHAR(255)
+                  constraints:
+                    nullable: false
+                    unique: true
```
no need in making it unique

jioter replied:
Removed

ahlukhouski2 commented on src/main/resources/db/changelog/init/01072021_create_country_table.yml:
```diff
@@ -0,0 +1,23 @@
+                    primaryKey: true
+                    primaryKeyName: pk_country_code
+                    nullable: false
+                    unique: true
```
no need in specifying unique here

jioter replied:
Removed

ahlukhouski2 commented on src/main/resources/db/changelog/init/sql/insert_data.sql:
```diff
@@ -4,37 +4,43 @@ VALUES (1, 'USER', current_timestamp, current_timestamp);
-INSERT INTO location (loc_id, loc_country, loc_city, loc_created, loc_modified)
-VALUES (6, 'Belarus', 'Minsk', current_timestamp, current_timestamp);
+INSERT INTO location (loc_id, country_code, loc_city,loc_address_line, loc_created, loc_modified)
+VALUES (4, 'Ukraine', 'Rivne', 'Beregova, 13', current_timestamp, current_timestamp);
```
countryCode should be used as a foreignKey, not countryFullName
same below

jioter replied:
Changed

ahlukhouski2 commented on src/main/resources/db/changelog/init/sql/insert_data.sql:
```diff
@@ -4,37 +4,43 @@ VALUES (1, 'USER', current_timestamp, current_timestamp);
+VALUES (10, 'Belarus', 'Mogilev', 'Aviatorov, 15', current_timestamp, current_timestamp);
+
+INSERT INTO location (loc_id, country_code, loc_city,loc_address_line, loc_created, loc_modified)
 VALUES (1, 'Sports and fitness', 'http://localhost/images/sport_image.jpeg', current_timestamp, current_timestamp);
```
country_code and loc_city, loc_address_line are missed here

jioter replied:
Changed

jioter committed:
Included Eq/Hcode to some fields
Removed unnessessary files/lines of code
Changed to functional style
Changed imput data in DB
Added transactional

jioter committed:
Removed native_query and used Spring Data JPA features instead

Added again Validation in VO, because it will add validation info to Swagger

jioter committed:
Add Search Specification to Location

jioter committed:
Deleted ond endpoints with search bby Country and City

jioter committed:
Fix Location Integration tests and location mapper

jioter committed:
Merge branch 'develop' of https://github.com/dmitriychernov7/sandbox-team3-be into location-lookup-table

 Conflicts:
	src/test/resources/sql/clean-up.sql
	src/test/resources/sql/location-init.sql

jioter committed:
Fix Location Integration tests and location mapper

Fix Controller Integration Tests

jioter committed:
Add Modif Location Controller Integration Test

jioter committed:
Merge branch 'develop' of https://github.com/dmitriychernov7/sandbox-team3-be into location-lookup-table

jioter committed:
Merge branch 'develop' of https://github.com/dmitriychernov7/sandbox-team3-be into location-lookup-table
